commcare-cloud/commcare_cloud/commcare_cloud.py

```diff
@@ -323,63 +343,121 @@ def make_parser(parser):
+                    include_vars = True
+                if become_user:
+                    cmd_parts += ('--become-user', args.become_user)
+                else:
+                    cmd_parts += ('--become',)
+
+            if include_vars:
+                cmd_parts += (
+                    '-e', '@{}'.format(os.path.expanduser('~/.commcare-cloud/vars/{env}/{env}_vault.yml'.format(env=args.environment))),
+                    '-e', '@{}'.format(os.path.expanduser('~/.commcare-cloud/vars/{env}/{env}_public.yml'.format(env=args.environment))),
+                )
+
+            ask_vault_pass = include_vars and public_vars.get('commcare_cloud_use_vault', True)
+            if ask_vault_pass:
+                cmd_parts += ('--vault-password-file=/bin/cat',)
+
+            cmd_parts += get_common_ssh_args(public_vars)
+            cmd = ' '.join(shlex_quote(arg) for arg in cmd_parts)
+            print(cmd)
+            p = subprocess.Popen(cmd, stdin=subprocess.PIPE, shell=True, env=ansible_context.env_vars)
+            if ask_vault_pass:
+                p.communicate(input='{}\n'.format(ansible_context.get_ansible_vault_password()))
+            else:
+                p.communicate()
+            return p.returncode
+
+        def run_check():
+            return _run_ansible(args, '--check', *unknown_args)
+
+        def run_apply():
+            return _run_ansible(args, *unknown_args)
+
+        exit_code = 0
+
+        if args.skip_check:
+            user_wants_to_apply = ask('Do you want to apply without running the check first?')
+        else:
+            exit_code = run_check()
+            if exit_code == 1:
+                # this means there was an error before ansible was able to start running
+                exit(exit_code)
+                return  # for IDE
+            elif exit_code == 0:
+                puts(colored.green(u"✓ Check completed with status code {}".format(exit_code)))
+                user_wants_to_apply = ask('Do you want to apply these changes?')
+            else:
+                puts(colored.red(u"✗ Check failed with status code {}".format(exit_code)))
+                user_wants_to_apply = ask('Do you want to try to apply these changes anyway?')
+
+        if user_wants_to_apply:
+            exit_code = run_apply()
+            if exit_code == 0:
+                puts(colored.green(u"✓ Apply completed with status code {}".format(exit_code)))
+            else:
+                puts(colored.red(u"✗ Apply failed with status code {}".format(exit_code)))
+
+        exit(exit_code)
+
+
+class RunShellCommand(RunAnsibleModule):
+    command = 'run-shell-command'
+    help = 'Run an arbitrary command via the shell module.'
+
+    @staticmethod
+    def make_parser(parser):
+        RunAnsibleModule._make_parser(
+            parser,
+            module_args_argument_name='shell_command',
+            module_args_argument_help="The shell command you want to run"
+        )
```

Comment:
So the way i solved this same problem in `AnsiblePlaybook` and the command classes that rely on it is to give `AnsiblePlaybook` one hard-coded interface, and then have the command classes that rely on it edit their own `args` and `unknown_args` variables to fit its interface. (For the shared parser args I use some good-enough intermediate base-class inheritance kind of thing, but I don't have a preference between that and something like what you did here with `_make_parser`.)

I like my edit-args-and-pass-on style because the central class needs to anticipate the needs of the commands that rely on it less, and there's less indirection code that needs to be in place to make it work. I'm happy to do this myself as a refactor later, though, if you're happy with that.